Update nav2_bringup readme for robot

mkhansen-intel opened the pull request:
This updates the readme with instructions for testing on a robot

mkhansen-intel commented:
Take another look. I've split the bringup into 2 parts and pushed 2 new launch files. Hopefully the README explains the new flow.

crdelsey commented on nav2_bringup/launch/nav2_bringup_1st_launch.py:
```diff
@@ -0,0 +1,32 @@
+            node_executable='amcl',
+            node_name='amcl',
+            output='screen',
+            parameters=[{ 'use_sim_time': use_sim_time}]),
```
We figured out that the `use_sim_time` parameter doesn't work, if I recall. You need to publish a parameter event after the node is up and listening.

We should just remove these and add another section to the readme saying run `ros2 param set /gazebo use_sim_time true`

mkhansen-intel replied:
@crdelsey - the use_sim_time does get set, but not for the transform, I've updated the readme to add that. Please look at the last commit.

orduno replied:
Works for AMCL, but I'm getting errors now for world_model. I need to trigger some param update for the warning to stop.

`WARN global_costmap: Costmap2DROS transform timeout. Current time: 1543355603.4700, global_pose stamp: 160.2020, tolerance: 0.3000, difference: 1543355443.2680`

mkhansen-intel commented:
OK, I updated per Carlos' comment. It now says to set use_sim_time for the world_model node also. That worked for me. I do find it frustrating that we have to do that every time but for now all we can do is document it.

mkhansen-intel commented:
Just rebased, let me know if more changes are needed

mkhansen-intel commented:
@SteveMacenski - I named them 1st and 2nd because they should preferrably be brought up in that order. The costmaps are in the 2nd one and they will wait indefinitely for the transforms to be available which are in the 1st one. Thus "bringup_1st" and "bringup_2nd". You 'can' use the nav2_bringup_launch.py file and just do it in one step, but it does make it harder to ensure that AMCL has localized correctly as the costmaps spew a ton of warning messages to the screen and can easily overwhelm the AMCL and map_server output.  Please feel free to give these a try and if you like I can make more changes. 

As for using the _launch.py instead of .launch.py there was a Discourse thread about this I believe a month or so back saying _launch should be the convention. 
https://discourse.ros.org/t/questioning-the-launch-py-file-extension/6352/2

SteveMacenski commented:
@mkhansen-intel to be explicit, you're referring to the error in this while loop? https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/src/costmap_2d_ros.cpp#L99 ?

I can resolve that issue later today if that's the blocker for having 1 launch file

orduno commented on nav2_bringup/README.md:
```diff
@@ -3,30 +3,68 @@
+
+Set the World Model node to use simulation time
+
+`ros2 param set /world_model use_sim_time True`
```
The global cost map doesn't get set. I still need to use `ros2 param set /global_costmap use_sim_time True`

mkhansen-intel replied:
I have not needed to set that one but I'm testing again now with latest master branch

mkhansen-intel replied:
OK, so I haven't seen exactly that /global_costmap needs to be set, but I have seen what I @crdelsey has also seen in that sometimes you have to set it twice. This whole setting of it is due to a bug in rclcpp I filed here:
https://github.com/ros2/rclcpp/issues/595

Anyway, until that is fixed, we have to set the use_sim_time variable after everything has been loaded. In some cases, it doesn't seem to take for all nodes, so we have to set it a 2nd time. 

In that case, we've seen you can simply set it again on /world_model and it should be OK. Again it's not perfect due to the bug above.

mkhansen-intel commented:
Update @SteveMacenski - Things are now stable enough that one launch file does seem to work most of the time, albeit with the caveat I just mentioned due to this bug:
https://github.com/ros2/rclcpp/issues/595

Sometimes we have to set the use_sim_time more than once before all the nodes respond to the event. 

Given that, I think the 'beginner' bringup should still be a 2-part bringup as I have here, so they can see AMCL localize correctly and make sure their transforms are working. 

However I will update the README again with more info stating this, and I will add the 'advanced' bringup of doing it all in one launch file. For those who have done the beginner and made that work, they could then move to the 1-part bringup after. Hopefully that should meet everyone's needs.

mkhansen-intel commented:
OK @SteveMacenski @orduno and @crdelsey - I made a few more changes to the README to hopefully make things as clear as they can be at this point in time. Please take a look at the latest.

Added notes and safety warnings
Added more details why use_sim_time must be set dynamically
Added 'Advanced bringup' instructions

SteveMacenski commented:
@mkhansen-intel I dont think its important for this PR to make it all 1 launch file. I'm trying to figure out the reasons why and how I (or someone) can potentially resolve them. 

Can you indicate if that's the line that's a problem for you or not?

mkhansen-intel commented:
@SteveMacenski - that line as well as the other messages from the Costmap 2D code that occur on every map update can overwhelm the screen output and make it hard to understand what is happening, until all the transforms are correct, AMCL receives valid scan data, and AMCL localizes. So I don't think you need to do anything, this is more of a way to help users get things correct on AMCL before proceeding. Once they have things working correctly using the 2 step flow, they can try the single step flow.

I think it would be great if you tried both and let me know what you think. I found that this helped me when we were debugging issues.
